### DISCUSSION TITLE
Headers already sent when errors are thrown from AsyncIterator

### DISCUSSION THREAD

**Discussion by bhamon** [enisdenjo/graphql-sse#96] [Q&A] [answered]
I've setup a basic test subscription that throws **inside** the async iterator:
```graphql
  type Subscription {
    test: Date!
  }
```

The subscribe function:
```js
import timers from 'node:timers/promises';
function resolverTest() {
  return (async function* () {
    const interval = timers.setInterval(1000);
    for await (const _ of interval) {
      yield {test: new Date()};
      throw new Error('test');
    }
  })();
}
```

Server setup:
I've copied it from the fastify TS file. The recipe from the website doc is not in sync and produce the same error but it was easier to debug with this one.
```js
import Fastify from 'fastify';
import { createHandler } from 'graphql-sse/lib/use/fastify';

const handler = createHandler({ schema });
const fastify = Fastify();
fastify.all('/graphql/stream', async (req, reply) => {
  try {
    await handler(req, reply);
  } catch (err) {
    console.error(err);
    reply.code(500).send();
  }
});

fastify.listen({ port: 4000 });
console.log('Listening to port 4000');
```

When I subscribe to `test` the server crashes with the following error:
`Error [ERR_HTTP_HEADERS_SENT]: Cannot write headers after they are sent to the client`

It comes from the `reply.code(500).send()`. If I remove this line the client hangs indefinitely.

Would it make sense to forward the error into the classic `errors` property of the response before closing the stream?
Or is there another way to signal the error now that the stream is already open?
Or I am missing something?

**Comment by enisdenjo** [comment, marked as answer]
Sorry for the delayed answer. The issue actually stems from [`graphql-js` not handling errors thrown during iteration](https://github.com/graphql/graphql-js/issues/4001). I'd prefer having this handled by the user (or graphql-js) instead.

I've [added](https://github.com/enisdenjo/graphql-sse/commit/718e57760b452431d9728804ac86548bb4129bdc) a recipe that showcases how you may achieve this by augmenting the `subscribe` function, [check it out here](https://the-guild.dev/graphql/sse/recipes#with-custom-subscribe-method-that-gracefully-handles-thrown-errors).

**Comment by bhamon** [comment]
No need to apology, we are all busy. Instead, thanks a lot for your time and for this amazing library.
I'll use your recipe in the meantime and wait for a proper support from `graphql-js`.
